### DISCUSSION TITLE
Offline cache/Sync? Plans for MongoDB support?

### DISCUSSION THREAD

**Discussion by p4-k4** [serverpod/serverpod#214] [Ideas]
A key component of firebase that is **missing** from other solutions such as [supabase](https://supabase.com) is how well it deals with offline caching firestore and maintaining sync.

Couldn't see anything in the roadmap but wondering whether there will be a default database solution *cough cough* mongodb and if there was, would there be any plans to support offline cache and server-client sync? I realise this is probably out of scope right now but thought I would start the conversation at least.

There's mongodb realm [flutter SDK](https://www.mongodb.com/docs/realm/sdk/flutter/) but it's in alpha which only supports a local database on the client. And is not a self-hosted solution.

**Comment by vlidholt** [comment]
This is a planned feature for the future.

The idea is to use the Postgres document database, which is as good or better than Mongo. An improvement over Firebase would be to also make it strictly typed.

Unfortunately, we won't be able to get this done for the 1.0 release, but it would be a really nice feature to add after that!

**Reply by Erik0806** [reply]
Now that 1.0 is out, do you still consider implementing it. I would really appreciate that

**Reply by vlidholt** [reply]
It will definitely happen. Unfortunately cannot give a timeline at this point though.

**Reply by alterhuman** [reply]
Any update on this?

**Reply by flutter-painter** [reply]
I would like to suggest a different approach to avoid offline lock- in and could help with this as I go along integrating serverpod.

In addition to Realm/Mongodb you could also mention Couchbase  **cbl-dart** which provide offline capacities.

**Reply by AndryHTC** [reply]
@vlidholt now that 1.2 is released... anything new on this matter? Would be another killer application

**Comment by kaljitism** [comment]
Hello @vlidholt , I am new to Backend and Server Side Programming. I am creating a Social Media application and Data Models for different classes is huge and complex for each instance and I want a lot of complex querying, an advanced search feature, using algorithms for recommendation system, etc. Would Postgre be a good and scaable choice for me??

**Reply by flutter-painter** [reply]
If I may, Postgre can do everything, BUT if you do not know yet how your models will evolve NoSQL is better suited.

For document DB in addition to mongoDB, you could also consider 
FaunaDB : https://pub.dev/packages/faunadb_http
Couchbase https://pub.dev/packages/couchbase

Also I wonder if graphQL could fit your complex querying needs
https://pub.dev/packages/graphql_flutter

**Reply by kaljitism** [reply]
Thanks for replying, I looked into these resources, they seem good but since I'm no expert on this, I can't comment if they will fulfill all my needs. However, I played with a few demo projects with Serverpod and they have a way to use dart for db queries which solves almost any scenario I could think of.

**Reply by AndryHTC** [reply]
@kaljitism I'm not sure what you mean with "complex" queries, but you are not forced to use the Serverpod ORM and the code generated models: you are free to use plain SQL that has been battle tested for decades.

Actually, SQL most of the time has more query capabilities than any NO-SQL DB. NO-SQL came out only for how easy it was to avoid strict schema + horizontally scale it, not for his powerful queries...

Also, Postgres is the most OSS used DB and has extensions for any use-care you like, even the latest kid on the block "Vector Embeddings".

If I was in your shoes I would care about query capabilities only if we talk about Document DB as Firestore or similar.

**Reply by kaljitism** [reply]
@AndryHTC  Thanks for addressing my issue. By complex querying, I mean - 

Every user in my app has close to 110 fields. Many are enums(some of which have 500+ values), Some are Strings, double, etc. Now, when a user is searching other users, I allow him to query all those fields about the other users. Which I belive could get pretty complex but I also believe I would be able to handle that with pure Dart. I am not sure if my skill level is on par to get same results with SQL. 

Since, I am new to Databases and I personally prefer querying through scripting rather than using a query language(coming from pandas and dataframe world!). I also researched a lot on different DBs and I think I'll be good with Postgres.
